### DISCUSSION TITLE
i want to change the font

### DISCUSSION THREAD

**Discussion by ProgramingHaven** [SethRobinson/proton#33] [Q&A]
how do i change the font without console crashing

**Comment by SethRobinson** [comment]
Info on [making new fonts is here](https://www.rtsoft.com/wiki/doku.php?id=proton:rtfont).

Or do you mean dynamically loading a different RTFont object on the fly to replace one earlier loaded?  I don't think I've ever tried that now that you mention it, I generally just load all the fonts at the same time and switch between them.

**Reply by iProgramMC** [reply]
Dynamic loading a different RTFont object over the existing one does not work on vanilla Proton. You have to modify `RTFont::Load` to kill the surface that already exists with `m_surf.Kill()` as follows:
```diff
	} else
	{
		m_fontStates.push_back(FontState('0', MAKE_RGB(255, 255, 255)));
	}

+	m_surf.Kill();

	//now load the actual bmp, which this pointer should be sitting at
	m_surf.SetTextureType(Surface::TYPE_GUI);

```

Once you add this, something like the following actually works fine. However you will need to rebuild your UI so that it doesn't look weird.

```cpp
// GetFont(FONT_SMALL) already loaded with a different font
GetFont(FONT_SMALL)->Load("interface/myfont.rtfont");
```


Regular assignment doesn't work, the new font object will use deleted surfaces if the old font object disappears because it does a regular memory copy on the RTFont instance instead of using an overloaded equals operator or copy constructor to copy the surface over.

**Reply by SethRobinson** [reply]
Huh, right after the SetTextureType, there is "if (!m_surf.LoadFileFromMemory(pSrcBytes))" and LoadFileFromMemory does have a Kill() at the start, so it seems like it should work as is, although the SetTextureType() will assert in debug mode due to a texture already being loaded...  Hrm.  I can add that Kill() though to avoid the assert, but it looks like it should work in release mode now actually.

Also, seems like you could do GetFont(FONT_SMALL)->GetSurface()->Kill() to manually kill the surface before the load.  I haven't tried any of this, just brainstorming. :)
